src/sema/eval.rs

```diff
@@ -196,3 +202,331 @@ pub fn eval_const_rational(
+                (&left, &right)
+            {
+                if right.is_zero() {
+                    Err(Diagnostic::error(loc, "divide by zero".to_string()))
+                } else {
+                    Ok(Expression::NumberLiteral(loc, ty, left % right))
+                }
+            } else {
+                Ok(Expression::Modulo(loc, ty, Box::new(left), Box::new(right)))
+            }
+        }
+        Expression::Power(loc, ty, unchecked, left, right) => {
+            let left = eval_constants_in_expression(*left)?;
+            let right = eval_constants_in_expression(*right)?;
+
+            if let (
+                Expression::NumberLiteral(_, _, left),
+                Expression::NumberLiteral(right_loc, _, right),
+            ) = (&left, &right)
+            {
+                if let Some(diagnostic) = overflow_check(right.clone(), Type::Uint(16), *right_loc)
+                {
+                    Err(diagnostic)
+                } else {
+                    Ok(Expression::NumberLiteral(
+                        loc,
+                        ty,
+                        left.pow(right.to_u16().unwrap().into()),
+                    ))
+                }
+            } else {
+                Ok(Expression::Power(
+                    loc,
+                    ty,
+                    unchecked,
+                    Box::new(left),
+                    Box::new(right),
+                ))
+            }
+        }
+        Expression::ShiftLeft(loc, ty, left, right) => {
+            let left = eval_constants_in_expression(*left)?;
+            let right = eval_constants_in_expression(*right)?;
+
+            if let (
+                Expression::NumberLiteral(_, _, left),
+                Expression::NumberLiteral(right_loc, _, right),
+            ) = (&left, &right)
+            {
+                if let Some(diagnostic) = overflow_check(right.clone(), Type::Uint(64), *right_loc)
+                {
+                    Err(diagnostic)
+                } else {
+                    Ok(Expression::NumberLiteral(
+                        loc,
+                        ty,
+                        left.shl(right.to_u64().unwrap()),
+                    ))
+                }
+            } else {
+                Ok(Expression::ShiftLeft(
+                    loc,
+                    ty,
+                    Box::new(left),
+                    Box::new(right),
+                ))
+            }
+        }
+
+        Expression::ShiftRight(loc, ty, left, right, bool) => {
+            let left = eval_constants_in_expression(*left)?;
+            let right = eval_constants_in_expression(*right)?;
+
+            if let (
+                Expression::NumberLiteral(_, _, left),
+                Expression::NumberLiteral(right_loc, _, right),
+            ) = (&left, &right)
+            {
+                if let Some(diagnostic) = overflow_check(right.clone(), Type::Uint(64), *right_loc)
```

Comment:
Just checking that the shift value fits into `uint64` is not enough.
```solidity
uint16 x = 1 >> 18446744073709551615;
```
This makes no sense. Please check that the shift value is less than `left.bits()`

Reply:
I followed the reasoning that the overall result shouldn't overflow. For instance, `int8 sesa = (1<<8) - 130;` doesn't throw an error in solc

Reply:
That is a good point.